### PR TITLE
Strict Execution

### PR DESCRIPTION
**Pull request by dmadisetti**
## 📝 Summary

Closes out #1142 and #1509 

This makes the kernel more deterministic by deep copying between cell passes, and ensuring reference existence before execution. It also restricts the scope of execution variables. Additionally, it introduces `mo.zero_copy` and `mo.shallow_copy`, which indicate to the kernel not to copy values (e.g. for very expensive variables).

Previous behavior:

image

New Behavior:

image

With zero_copy

image

Missing:
  - [x] More unit tests (specifically ast / executor)
  - [ ] UI love for Error on missing references
  - [x] Documentation

## 🔍 Description of Changes

This ended up being a much bigger change than expected. I think it still needs some more unit tests, but wanted to get some eyes on it / see if it passes  CI before doing anything more.

It's a pretty invasive change adding an "Executor" class and a copy module. Also had to get block level refs, so it touches AST. Also had to get state working, so had to tinker with that. Took some license with the structuring.

Execution is under "experimental" since I think this needs to be used a little before stability guarantees. currently it does not support script or direct cell runs.

## 📜 Reviewers

@akshayka OR @mscolnick

### CODE REVIEW

File: marimo/_runtime/executor.py
Comment (dmadisetti) [inline marimo/_runtime/executor.py:221]: "dictionaty"

**Review by dmadisetti — COMMENTED** [review]
Will require some clean-up iterations- but I'm going to go enjoy my Saturday!
Sorry for the huge PR

**Comment by dmadisetti** [comment]
??? Unsure why __future__ isn't working for 3.8, will investigate later

**Comment by akshayka** [comment]
@dmadisetti This is a very interesting PR, and the implementation is quite clean. 

I have one concern, though — this change will put marimo in a situation where some notebooks only work when execution mode is strict, whereas others will only work when execution mode is relaxed. This would make sharing marimo notebooks more difficult, since users would need to somehow know which runtime configuration to use. 

For this reason I'm not fully convinced that the benefits are large enough to warrant the complexity. But it's possible I am missing something ... Beyond the matplotlib example, how do you envision this feature being used? 

An alternative we could consider: perhaps this feature could be a good motivating candidate for the extension system we talked about.

I know we discussed this feature in #1142, but I was under the impression that that discussion was for `app.run()` and exports, not for the core notebook experience.

**Comment by dmadisetti** [comment]
All relaxed notebooks should be able to run strict notebooks, but not all strict notebooks will be able to run relaxed ones (maybe there's an example I'm not seeing?).

In the cases where notebook behavior is different, in the "relaxed" case- it's indeterminant (For instance the plot example, I'm not sure why $x^3$ ran before $x^2$). Strict mode can still force "relaxed" behavior with zero_copy, and a strict mode user should take that into account if writing a notebook for a general audience. Being opt-in, comes with the understanding that the notebook must be written more carefully

I think this has come from some of my frustrations that marimo seems so close- and then I'll get an unexpected mutation- at which point it's just easier to restart the kernel. This solves that issue.

I've been trying to write some of my code with this kernel, and it's been interesting. I'm catching bugs I don't think I normally would with marimo- I even found a marimo bug with regard to exceptions #1571. It's just a step up wrt reproducibility

But yeah, it is a little invasive, but I don't think there's much overhead to normal operations. If it's too different, it could be the basis for tokarip

**Comment by akshayka** [comment]
Thanks for explaining; the step up  in reproducibility makes sense, and it's cool you were able to find a bug in marimo using this.

There are cases in which I've used mutations when making UIs. There was no indeterminism there (though of course the notebook was stateful), but those notebooks were written before we had mo.state.

The RAM overhead could be very large when working with numerical data, but I suppose since it's opt-in the user knows what they're signing up for.

This is definitely useful ... Let me think a little bit more about whether this goes in marimo core (perhaps it could), or if it's the basis for tokarip?

**Comment by dmadisetti** [comment]
For UI and state, there's a check that makes them automatically "no copy". but it won't catch cases like:

```python
states = [State(), State()]
```

since these are nested in a data structure.

So **1)** the strict kernel could reimplement deepcopy to do the type check _or_ **2)** stay as is, and much more simply, we can make State, UI Elements, and virtual files not deepcopyable to catch these edge cases, which seems correct beyond this PR (I don't think copy falls under their intended behavior as is—for example, I found that the ref counter with virtual files breaks if the file variable is copied).

However, you can still get the same behavior with:

```python
states = mo.copy.zero_copy([State(), State()])
# or even states = mo.copy.shallow_copy([State(), State()]) in this case
```

`marimo._ast` objects are also whitelisted. LMK if you can think of any others.

As is, I would not recommend making this as anything but experimental until it has had some actual use. I keep finding weird cases like "calling a lambda defined in another cell but using a private variable in scope and another variable in defined in a 3rd cell" will fail (this actually passes now, but just as an example).

Tests are passing after adding `strict_kernel` to `all_kernels`, and I have a local commit that adds the runtime/ state tests to be run with `strict_kernel` (there's one failing case based on `required_refs` resolution that I want to fix before I push)

Re memory overhead: This will never be more than x2 (the reference copy and the working copy), with the idle state being the same amount of memory. But this is also why I added `zero_copy`, since some data is too big to reasonably be copied every reference.

But yes, this could feasibly be extracted to something like tokarip, if the visitor changes and execution register stay

---

I don't know this is obviously a little opinionated, maybe we should find some willing beta testers

**Comment by dmadisetti** [comment]
Hmm. Weird. Failing with timeout on 3.11 which is also suspiciously slow with trivial changes too: https://github.com/marimo-team/marimo/actions/runs/9437984531/job/25994902883?pr=1583

**Comment by akshayka** [comment]
> Failing with timeout on 3.11 which is also suspiciously slow with trivial changes too:

Timeout is okay, 3.11 runs code coverage, so sometimes times out.

**Comment by dmadisetti** [comment]
Quickly ran through the tutorials. They all work! Except this by design

image

File: marimo/_ast/visitor.py
Comment (akshayka) [inline marimo/_ast/visitor.py:34]: Oh, this is interesting -- I was just thinking how it would be useful to have this information for functions and classes. Is it correct to say that if `required_refs` is the empty set, then the function or class is pure?

File: marimo/_ast/visitor.py
Comment (akshayka) [inline marimo/_ast/visitor.py:82]: Can you add a comment describing the semantics of `ref_stack`?

File: marimo/_ast/visitor.py
Comment (akshayka) [inline marimo/_ast/visitor.py:295]: I kind of understand what's happening here. But I don't understand the point of 

```python
self.ref_stack[-1].update(refs)
```

I know I'm just misunderstanding something -- can you add a comment explaining what's happening?

File: marimo/__init__.py
Comment (akshayka) [inline marimo/__init__.py:98]: Let's keep `copy` private for now, until we have more confidence in the executor and its associated APIs

File: tests/_ast/test_visitor.py
Comment (akshayka) [inline tests/_ast/test_visitor.py:373]: Yes, I think they are required refs -- can't define `foo` without them, right?

**Review by akshayka — COMMENTED** [review]
Not a full code review, but left a few comments. I will continue to review later.

Overall, I think this is a great PR with several useful new features. I'm still not sure what the best way to expose the strict runtime to users will be, but like you've mentioned, we can punt on that and keep it experimental for now.

I actually think the way you've architected things, this PR isn't invasive. The executor abstraction is clean and not entangled with the rest of the codebase; easy to swap in and out.

File: marimo/_ast/visitor.py
Comment (dmadisetti) [reply marimo/_ast/visitor.py:34]: It's not guaranteed to be pure since you might pass in a file descriptor or do something like `__import__("requests")`, but it's a good indication that the function _might_ be pure.

File: tests/_ast/test_visitor.py
Comment (dmadisetti) [reply tests/_ast/test_visitor.py:373]: Just checked, they are not: https://marimo.app/l/7swcsf

I'll take a look - but not sure it's worth trying to tease out this edgecase

File: marimo/_ast/visitor.py
Comment (dmadisetti) [reply marimo/_ast/visitor.py:91]: Added a comment here, but even more detail under variabledata.required_refs

File: tests/_ast/test_visitor.py
Comment (akshayka) [reply tests/_ast/test_visitor.py:373]: `del` just decrements the ref count though — `foo` still refers to the `Namespace` object, for example:

image

But I guess the symbols aren't required to be in globals, but instead are bound at function construction time.

Anyway I think it's fine as is

**Comment by dmadisetti** [comment]
Cool! Good comments. I hope my documentation isn't too overwrought. Today is the first day I used this branch without actively tinkering with it so

Notes from usage:
 - [ ] `NameError` catching is a little too aggressive right now. I was developing from a module and the way the error is currently thrown obfuscates the stack trace in a way that was not ideal.
 - [ ] `StopError` wasn't producing the correct UI behavior? Same area of code that needs refactor / checking
 
Misc todos
 - [ ] Clean up type checking in execution—currently, there are a bunch of random imports. I think this can be done more elegantly.
 - [ ] investigate `required_refs` from function args
 - [ ] (maybe) expand required ref precedent to ignore `np.ndarray[number]`, and make it easily extensible for more types.

 A different PR/ needs discussion:
  - [ ] prevent unintentional clone of state, files, and UI objects. These objects aren't meant to be cloned and might lead to unexpected behavior if they are cloned- so by either throwing an error in `__copy__` or `__reduce__` (or just removing the relevant magic functions), we can stop inadvertent improper usage.

I think unit tests are OK- and I'm going to hold off for external documentation until/ if this is exposed.

File: marimo/_runtime/executor.py
Comment (akshayka) [inline marimo/_runtime/executor.py:309]: Can you explain what's happening here with a comment? I don't follow.

File: marimo/_runtime/executor.py
Comment (akshayka) [inline marimo/_runtime/executor.py:292]: When would this happen? My understanding is `lcls` are the variables the cell defined after running, and `glbls` are the variables that were defined before it ran.

File: marimo/_runtime/executor.py
Comment (akshayka) [inline marimo/_runtime/executor.py:295]: Should this be moved to a utility function? There's an `is_local` function in `visitor`, for example.

File: marimo/_runtime/executor.py
Comment (akshayka) [inline marimo/_runtime/executor.py:339]: Wow, very thorough — nice.

File: marimo/_runtime/executor.py
Comment (akshayka) [inline marimo/_runtime/executor.py:314]: Can you add one sentence somewhere that describes what the returned callable does?

Something like: Returns a callable that takes a variable name and its associated data, and returns `True`  if ...

File: marimo/_runtime/executor.py
Comment (akshayka) [inline marimo/_runtime/executor.py:379]: Fancy — are you doing this so the module doesn't have to be installed?

**Review by akshayka — COMMENTED** [review]
Small comments/questions

**Comment by akshayka** [comment]
All tests passing! 👀 🎉

File: marimo/_runtime/executor.py
Comment (dmadisetti) [reply marimo/_runtime/executor.py:379]: That's the intent; not certain if it works right now

I think the last change I want to make is to test/get this working since it seemed like it wasn't catching some things. That will probably have to wait for tomorrow evening.

File: marimo/_runtime/executor.py
Comment (dmadisetti) [reply marimo/_runtime/executor.py]: Moved, but in retrospect maybe `marimo.ast.variables` might have been better and moved over the data classes too.

File: marimo/_runtime/executor.py
Comment (dmadisetti) [reply marimo/_runtime/executor.py:381]: Actually, wasn't that bad. pushed this too

File: marimo/_runtime/executor.py
Comment (dmadisetti) [reply marimo/_runtime/executor.py:293]: Added a comment, `glbs` still may have a stale variable in place- and cells may depend on that stale variable

File: marimo/_runtime/executor.py
Comment (dmadisetti) [reply marimo/_runtime/executor.py]: The intent was that a private variable that was never used, would not get put into globals. But an unused private variable can still be an eval output, and potentially some other unexpected cases- good catch.

**Comment by dmadisetti** [comment]
More misc:
 - [ ] revisit exceptions

image

---

Maybe over the weekend, just noticed this while working

File: marimo/_runtime/copy.py
Comment (akshayka) [inline marimo/_runtime/copy.py:56]: Is the point of this to try to prevent mutations of attributes? My understanding is that re-assignment will fail, but mutations by calling methods on attributes will still work.

File: marimo/_runtime/copy.py
Comment (akshayka) [inline marimo/_runtime/copy.py:81]: Can you add a comment explaining why some dunder attrs are being excluded from slots?

File: marimo/_runtime/copy.py
Comment (akshayka) [inline marimo/_runtime/copy.py:70]: Does that mean a zero-copied NumPy array could fail if involved in a binary operation with a vanillia NumPy array?

**Review by akshayka — COMMENTED** [review]
Few more questions

File: marimo/_runtime/copy.py
Comment (dmadisetti) [reply marimo/_runtime/copy.py:56]: This is the closest I could get to a "move"

Idea is we check the variable type for ZeroCopy or ShallowCopy to determine clone behavior. Preventing direct mutation is a needed consequence of exposing all the attributes.

We could also not expose the attributes and always require an unwrap or call, and not do this "shadowing"

Comment (dmadisetti) [reply marimo/_runtime/copy.py:56]: Zero copy could just put the ID in a special dictionary. Much less complicated, but does require some trickery 

I think this is a better idea, just type checking was my first idea. Thoughts?

File: marimo/_runtime/copy.py
Comment (dmadisetti) [reply marimo/_runtime/copy.py:70]: So in numpy

'B + A' works but not 'A + B'

File: marimo/_runtime/copy.py
Comment (dmadisetti) [reply marimo/_runtime/copy.py:56]: Type checking does have the consequence of being conscious of your global variables though, and the RO forces you to unwrap and makes it clear state change is happening.

Particular to numpy, there is `.view()` which can called.

File: marimo/_runtime/copy.py
Comment (akshayka) [inline marimo/_runtime/copy.py:30]: `ReadOnlyError`, for readability?

File: marimo/_runtime/copy.py
Comment (akshayka) [reply marimo/_runtime/copy.py:56]: > This is the closest I could get to a "move"

As in `std::move` from C++? I'm not sure I see the connection, maybe I'm missing something obvious

> Zero copy could just put the ID in a special dictionary. Much less complicated, but does require some trickery

And then get the original object in subsequent cells? That might be simpler, and would fix the issue with associative operators that you mentioned. It would also be compatible with any code that relied on checking the `type` of a zero copied object, whereas wrapping doesn't appear to be:

image

Could a special dictionary be used for `ShallowCopy`, too?

File: marimo/_runtime/copy.py
Comment (akshayka) [inline marimo/_runtime/copy.py:45]: I suppose this would go in the docs you'll write in a future PR (no need right now). But out of curiosity, when would someone choose `shallow_copy` over `zero_copy`?

File: marimo/_runtime/copy.py
Comment (akshayka) [reply marimo/_runtime/copy.py:56]: > Type checking does have the consequence of being conscious of your global variables though, and the RO forces you to unwrap and makes it clear state change is happening.

By the time you're using strict and opting into things like `zero_copy` — I imagine you're already being pretty conscious of your global variable usage. 

> the RO forces you to unwrap and makes it clear state change is happening

That's true.

I would opt for a special dictionary because it's simpler and unlike type checking doesn't have gotchas when interacting with the "wrapped" object. But I'll leave it up to you to determine the benefits of typing outweigh its costs.

File: marimo/_runtime/copy.py
Comment (dmadisetti) [reply marimo/_runtime/copy.py:45]: A container of object that's shouldnt be copied, like `[state, state, state]`

File: marimo/_runtime/copy.py
Comment (dmadisetti) [reply marimo/_runtime/copy.py:56]: Implementation wise, there would just be a special dictionary for zero and shallow copy, that the helper functions would just interact with, and then a check would be made from the global objects. Invalidating IDs on rerun might be tricky. Also, Python reuses IDs. But if not going down the id route, then there might be some manual gc that would have to happen 

I don't think it would be hard to put in but maybe a few edge cases- so I'm inclined to leave it as is for now

Comment (akshayka) [reply marimo/_runtime/copy.py:56]: > Implementation wise, there would just be a special dictionary for zero and shallow copy, that the helper functions would just interact with, and then a check would be made from the global objects. Invalidating IDs on rerun might be tricky. Also, Python reuses IDs. But if not going down the id route, then there might be some manual gc that would have to happen

Yea. That makes sense to me. I had to solve related edge cases for our registry of UI elements, during which process I learned that Python reuses IDs.

> I don't think it would be hard to put in but maybe a few edge cases- so I'm inclined to leave it as is for now

Ok sounds good.

File: marimo/_runtime/copy.py
Comment (akshayka) [reply marimo/_runtime/copy.py:45]: > A container of object that's shouldnt be copied, like [state, state, state]

If you don't want something to be copied, wouldn't you use `zero_copy`, not `shallow_copy`?

Comment (dmadisetti) [reply marimo/_runtime/copy.py:45]: The container can mutate w/ ZeroCopy but not shallow copy

`states.append(state)`

will be reset w/ shallow copy

**Comment by dmadisetti** [comment]
Exceptions now look good 
image

**Comment by dmadisetti** [comment]
Sorry for the massive PR. Even though most of it is comments and tests, I think I could have broken it up.

Will try to provide quicker turn around for additional comments since I know this is now blocking

**Comment by mscolnick** [comment]
(its not blocking, I can still work off my branch) thanks for the effort on this one! definitely a tricky problem to solve

**Review by akshayka — APPROVED** [review]
LGTM! At this point, since it's experimental -- I think it's good to merge. I'm sure we'll learn more as you and others try it out.

Thanks for the PR!

**Pull request merged**
